### DISCUSSION TITLE
GetResourceAttribution() assertion

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#217] [Q&A]
I'm getting a debug assertion exception when trying to create a graphics pipeline state. It looks like the assertion is being triggered by a missing/blank resource name? I'm having trouble figuring out exactly what's wrong? Any guesses?

![exception_01](https://user-images.githubusercontent.com/43590381/200209725-6b40bd36-bcc4-4e02-aba5-2be2401c54ff.png)

Let me know if you need more information. Thanks!

**Comment by TheMostDiligent** [comment]
It is not clear from your screenshot which exactly assertion was triggered - the line number is cut off.

**Comment by RobertMtx** [comment]
Sorry, it's line 969, which I believe is the first line of code in `GetResourceAttribution()`:
` VERIFY_EXPR(Name != nullptr && Name[0] != '\0');`

I tried to follow the logic path to see what name it is referring to, but haven't been able to figure it out. But the watch shows the name is empty, whatever it is.

When I follow it back, it looks like its being triggered for the first constant buffer of my resource signature. But I definitely gave the constant buffer a name in the resource signature. Would there be some other name string I've missed?
```
    void ProcessResources(THandleCB          HandleCB,
                          THandleSampler     HandleSampler,
                          THandleTexSRV      HandleTexSRV,
                          THandleTexUAV      HandleTexUAV,
                          THandleBufSRV      HandleBufSRV,
                          THandleBufUAV      HandleBufUAV,
                          THandleAccelStruct HandleAccelStruct) const
    {
        for (Uint32 n = 0; n < GetNumCBs(); ++n)
        {
            const auto& CB = GetCB(n);
            HandleCB(CB, n);
        }

```

**Reply by TheMostDiligent** [reply]
Can you post your shaders or at least resources that they use?

**Comment by RobertMtx** [comment]
Sure. The shaders are pretty simple, but I use typename aliases in my C++ code for some of the Diligent typenames that are global. Usually those start with grx. Let me know if you need clarification for anything. Do you think its something I'm doing wrong in the shader code itself? The shader files I included here are actually several files that were pre-processed down into a single file before being sent to the compiler. I'm using the DXC compiler at runtime to generate the byte-code, then sending that to Diligent to create the pipeline state. Let me know if you need to see the code related to the pipeline state. As far as resources, I just link to this resource signature (*.h in the zip).

I really appreciate any advice, if you have any ideas on what I may be doing wrong,

[Temp.zip](https://github.com/DiligentGraphics/DiligentEngine/files/9953217/Temp.zip)

Here is some of the data related to those global buffers in the resource signature. I double checked the name strings, and they are all good at the time of creating the pipeline state.
```
constexpr SXATTRSPEC _SxAttrSpecTable[int(sxAttr::COUNT)] = {

    // special NONE slot
    SXATTRSPEC( sxAttr::NONE,       EMPTY,          0,                      0,      sxFreq::NONE,   gpuOp::none,                                gpuBind::none ),

    // vertex buffer instance indices
    SXATTRSPEC( sxAttr::INSTANCE,   EMPTY,          sizeof(sxrSlot),        2048,   sxFreq::INSTX,  gpuOp::mSTANDARD | gpuOp::uSTATIC,          gpuBind::VERTEX_BUFFER ), // created at startup only - max capacity represents the max number of elements we can draw in a single cycle (requiring more cycles)
    
    // index quad buffer indices
    SXATTRSPEC( sxAttr::QUAD,       EMPTY,          sizeof(gpuIdx),         1536,   sxFreq::QFACE,  gpuOp::mSTANDARD | gpuOp::uSTATIC,          gpuBind::INDEX_BUFFER ), // rendered index buffer
    
    // drawing cache buffer
    SXATTRSPEC( sxAttr::CACHE,      EMPTY,          sizeof(sxvBake),        1024,   sxFreq::CACHE,  gpuOp::mSTANDARD | gpuOp::uDYNAMIC,         gpuBind::VERTEX_BUFFER ), // drawing cache buffer

    // global uniform constant buffer
    SXATTRSPEC( sxAttr::SCENE,      "Scene",        sizeof(sxattrScene),    1,      sxFreq::FRAME,  gpuOp::mSTRUCTURE | gpuOp::uMUTABLE,        gpuBind::UNIFORM_BUFFER ), // mutable for heavy GPU access and less CPU changes

    // environment data, such as lighting and environment map indices
    SXATTRSPEC( sxAttr::ENVSETS,    "bufEnv",       sizeof(sxattrEnvSet),   64,     sxFreq::FRAME,  gpuOp::mSTRUCTURE | gpuOp::uDYNAMIC,        gpuBind::SHADER_RESOURCE ), /// TODO: change to MUTABLE??

    // buffers containing simple values & custom structures
    SXATTRSPEC( sxAttr::VECPAL,     "bufVec",       sizeof(sxattrVector),   1024,   sxFreq::CYCLE,  gpuOp::mSTRUCTURE | gpuOp::uDYNAMIC,        gpuBind::SHADER_RESOURCE ),
    SXATTRSPEC( sxAttr::COORDPAL,   "bufCoord",     sizeof(sxattrCoords),   1024,   sxFreq::CYCLE,  gpuOp::mSTRUCTURE | gpuOp::uDYNAMIC,        gpuBind::SHADER_RESOURCE ),
    SXATTRSPEC( sxAttr::XFORMPAL,   "bufXForm",     sizeof(sxattrXForm),    1024,   sxFreq::CYCLE,  gpuOp::mSTRUCTURE | gpuOp::uDYNAMIC,        gpuBind::SHADER_RESOURCE ),
    SXATTRSPEC( sxAttr::WORLD,      "bufWorld",     sizeof(sxattrWorld),    512,    sxFreq::CYCLE,  gpuOp::mSTRUCTURE | gpuOp::uDYNAMIC,        gpuBind::SHADER_RESOURCE )
};
```

**Reply by TheMostDiligent** [reply]
If you use DXC, make sure that you don't strip reflection information from the shaders.

**Reply by RobertMtx** [reply]
That might be exactly what it is then. I think they recommended stripping everything from the shaders. I'll try to make those changes and test again. Thanks!

Do you have any recommended compiling args to send to it? Below is what I am currently sending. I just commented out the reflection strip line.
```
// entry point
Args.AddParm( L"-E" );
Args.AddParm( EntryDefault );
ArgEntry = Args.SlotLast();

// shader model
Args.AddParm( L"-T" );
Args.AddParm( L"vs_x_x" );
ArgModel = Args.SlotLast();

// generate symbols
Args.AddParm( L"-Zi" );
Args.AddParm( L"-Fd" );
Args.AddParm( PdbPath );

// strip data from shader object
// Args.AddParm( L"-Qstrip_reflect" );
Args.AddParm( L"-Qstrip_debug" );
Args.AddParm( L"-Qstrip_priv" );
Args.AddParm( L"-Qstrip_rootsignature" );
```
Note that I programmatically change the model arg to vs_6_0 or ps_6_0 right before compiling.

**Reply by RobertMtx** [reply]
Yes, the stripped reflection data was the reason for the missing resource name. Now on to the next problem, haha.

Thanks, I appreciate the assist! You may want to consider adding a message about missing reflection data when that name string is blank, for future cases of programmers like me, half blindly falling down the rabbit hole. Thanks again!

**Reply by TheMostDiligent** [reply]
> You may want to consider adding a message about missing reflection data when that name string is blank

Yes, that is a good suggestion.

**Comment by RobertMtx** [comment]
I have another question about the texture array situation, but didn't want to create another thread, since I'm hoping the answer is probably a simple one. I'm not sure if you remember, but I'm defining a texture array of 255 slots in the shader, but only using the slots that are "registered" at any one time. But Diligent is giving me an error  "_Diligent Engine: ERROR: No resource is bound to variable 'texReg[50]' in shader '0000' of PSO '0000'._", for each slot that currently doesn't have a texture assigned to it. Should I be assigning some dummy texture to all unused slots? Or is there perhaps a better way to handle that situation?

**Reply by TheMostDiligent** [reply]
Diligent can't know which textures will be used by the shader. Assigning some texture should fix this.
